### PR TITLE
[JENKINS-50336] Fix loading optional extensions after installing a missing dependency

### PR DESCRIPTION
**Pull request by Vlatombe**
See [JENKINS-50336](https://issues.jenkins-ci.org/browse/JENKINS-50336).

### Proposed changelog entries

* Optional extensions are now loaded without requiring to restart Jenkins after installing an optional dependency

### Submitter checklist

- [X] JIRA issue is well described
- [X] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [X] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@reviewbybees

### CODE REVIEW

**Review by kuisathaverat — APPROVED** [review]
🐝

File: test/src/test/java/hudson/PluginManagerTest.java
Comment (daniel-beck) [inline test/src/test/java/hudson/PluginManagerTest.java]: So this does not actually test the much more common case of `@Extension(optional = true)`?

**Comment by Vlatombe** [comment]
@jenkinsci/code-reviewers anyone interested to review this one?

**Comment by oleg-nenashev** [comment]
Please do not merge right now, we may need out of order release after the
regression in the security patch

File: test/src/test/java/hudson/PluginManagerTest.java
Comment (dwnusbaum) [reply test/src/test/java/hudson/PluginManagerTest.java]: Yes, but I do not have a demo or anything at this point. I'm not sure what the best way would be to handle custom code in the mock plugins, since writing java code in strings seems less than ideal, but I guess for a case like this it would be totally fine.

**Review by jglick — COMMENTED** [review]
Leaving this for @dwnusbaum to evaluate.

**Review by dwnusbaum — APPROVED** [review]
I tested locally and the change looks good to me.

**Review by oleg-nenashev — APPROVED** [review]
LGTM

**Pull request merged**
